rpc_proxy.go

```diff
@@ -0,0 +1,398 @@
+		// gRPC web requests to have it and "convert" the auth into a
+		// proper macaroon now.
+		ctx, err := p.basicAuthToMacaroon(ss.Context(), info.FullMethod)
+		if err != nil {
+			return fmt.Errorf("error upgrading basic auth: %v", err)
+		}
+
+		// With the basic auth converted to a macaroon if necessary,
+		// let's now validate the macaroon.
+		err = p.macValidator.ValidateMacaroon(
+			ctx, uriPermissions, info.FullMethod,
+		)
+		if err != nil {
+			return err
+		}
+
+		return handler(srv, ss)
+	}
+}
+
+// basicAuthToMacaroon checks that the incoming request context has the expected
+// and valid basic authentication header then attaches the correct macaroon to
+// the context so it can be forwarded to the actual gRPC server.
+func (p *rpcProxy) basicAuthToMacaroon(ctx context.Context,
+	requestURI string) (context.Context, error) {
+
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return ctx, nil
+	}
+
+	authHeaders := md.Get("authorization")
+	if len(authHeaders) == 0 {
+		// No basic auth provided, we don't add a macaroon and let the
+		// gRPC security interceptor reject the request.
+		return ctx, nil
+	}
+
+	// The user specified an authorization header so this is very likely a
+	// gRPC Web call from the UI. But we only attach the macaroon if the
+	// auth is correct. That way an attacker doesn't know that basic auth
+	// is even allowed as the error message will only be the macaroon error
+	// from the lnd backend.
+	authHeaderParts := strings.Split(authHeaders[0], " ")
+	if len(authHeaderParts) != 2 {
+		return ctx, nil
+	}
+	if authHeaderParts[1] != p.basicAuth {
+		return ctx, nil
+	}
+
+	var (
+		macPath string
+		err     error
+	)
+	switch {
+	case isLndURI(requestURI):
+		_, _, _, macPath, err = p.cfg.lndConnectParams()
+		macPath = path.Join(macPath, defaultLndMacaroon)
+
+	case isLoopURI(requestURI):
+		macPath = p.cfg.Loop.MacaroonPath
+
+	case isFaradayURI(requestURI):
+		macPath = p.cfg.Faraday.MacaroonPath
+
+	default:
+		return ctx, fmt.Errorf("unknown gRPC web request: %v",
+			requestURI)
+	}
+	if err != nil {
+		return ctx, fmt.Errorf("error getting macaroon path: %v", err)
+	}
+
+	// Now that we know which macaroon to load, do it and attach it to the
+	// request context.
+	macBytes, err := readMacaroon(macPath)
+	if err != nil {
+		return ctx, fmt.Errorf("error reading macaroon: %v", err)
+	}
+	md.Set(HeaderMacaroon, hex.EncodeToString(macBytes))
+	return metadata.NewIncomingContext(ctx, md), nil
+}
+
+// dialLnd connects to lnd through the given address and uses the given TLS
+// certificate to authenticate the connection.
+func dialLnd(dialAddr, tlsCertPath string) (*grpc.ClientConn, error) {
+
+	var opts []grpc.DialOption
+	tlsConfig, err := credentials.NewClientTLSFromFile(tlsCertPath, "")
+	if err != nil {
+		return nil, fmt.Errorf("could not read lnd TLS cert %s: %v",
+			tlsCertPath, err)
+	}
+
+	opts = append(
+		opts,
+		grpc.WithCodec(grpcProxy.Codec()), // nolint
```

Comment:
why do we need `WithCodec`?

Reply:
According to the `grpcProxy` godoc, this is crucial for the gRPC web -> gRPC native proxy to work. Added a comment to clarify.